File: grokking_algorithms/jkoers/ch01_introduction/ts/main.ts
import fs from 'fs'

function bin_search_i(arr: any[], target: any): number {
	let low = 0
	let high = arr.length - 1

	while (low <= high) {
		const mid = Math.floor((low + high) / 2)
		if (arr[mid] == target)
			return mid
		else if (arr[mid] < target)
			low = mid + 1
		else
			high = mid - 1
	}
	return -1
}

function measure(names: any[], find: any) {
	console.time('Search took')
	const index: number = bin_search_i(names, find)
	console.timeEnd('Search took')

	if (index < 0)
		console.log(`${find} not found`)
	else
		console.log(`Found ${find} at index ${index}`)
}

if (process.argv.length != 4) {
	console.log(`Usage: ${process.argv[0]} ${process.argv[1]} <input>`)
	process.exit(1)
}

const filename: string = process.argv[2]!
const find: string = process.argv[3]!
const names: string[] = fs.readFileSync(filename).toString().split('\n').sort()

console.log(`Searching for ${find} in array of ${names.length}...`)

for (let i = 0; i < 10; i++)
	measure(names, find)

File: grokking_algorithms/jkoers/ch01_introduction/cpp/main.cpp
#include "main.hpp"

#include <algorithm>
#include <chrono>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

std::vector<std::string> get_names(const std::string& filename, size_t max_names = ~0) {
	std::ifstream			 file(filename);
	std::vector<std::string> names;

	while (!file.eof() || !max_names--) {
		std::string line;
		getline(file, line);
		names.emplace_back(line);
	}
	std::sort(names.begin(), names.end());
	return (names);
}

int main(int argc, char** argv) {
	if (argc != 3) {
		std::cerr << "Usage: " << argv[0] << " <filename> <find>" << std::endl;
		return (EXIT_FAILURE);
	}
	const std::string			   filename = argv[1];
	const std::string			   find = argv[2];
	const std::vector<std::string> names = get_names(filename);

	std::cout << "Searching " << find << " in array of size " << names.size() << "..." << std::endl;

	const auto	  start = std::chrono::high_resolution_clock::now();
	const ssize_t i = bin_search_i(names, find);
	const auto	  end = std::chrono::high_resolution_clock::now();
	const auto	  duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);

	if (i < 0)
		std::cout << "Not found" << std::endl;
	else
		std::cout << "Found at position " << i << " in " << duration.count() << " microseconds" << std::endl;

	return (EXIT_SUCCESS);
}
